### DISCUSSION TITLE
Can I move forward with ondemand parsing if error occurs?

### DISCUSSION THREAD

**Discussion by kgs** [simdjson/simdjson#1931] [Q&A] [answered]
Hi, I wonder if it is possible to skip errors and move forward with parsing if some error occurs (in variant without exceptions).
For example I expect string, but got int. I just want to skip this field and go to next one.

I would also like to know if it is possible to move forward (error is not critical), or this is not possible because JSON structure is broken for example (in DOM parser this was possible since validation was at the beginning).

Like with this example ("make" should be string).
```C++
    ondemand::parser parser;
    auto car_json = R"(
        {
            "make": 42,
            "model": "Camry,
            "year": 2018,
            "tire_pressure": [ 40.1, 39.9, 37.7, 40.4 ]
         }
    )"_padded;
    ondemand::document doc;

    auto error = parser.iterate(car_json).get(doc);
    if (error) {
        std::cerr << error << std::endl;
        return false;
    }

    ondemand::object car; // invalid until the get() succeeds
    error = doc.get_object().get(car);
    if (error) {
        std::cerr << error << std::endl;
        return false;
    }

    // Accessing a field by name
    std::string_view make;
    std::string_view model;
    error = car["make"].get(make);
    if (error) {
        std::cerr << error << std::endl;
        // DO NOT RETURN
    }
    error = car["model"].get(model);
    if (error) {
        std::cerr << error << std::endl;
        // DO NOT RETURN
    }
    ondemand::array pressures;
    error = car["tire_pressure"].get_array().get(pressures);
    if (error) {
        std::cerr << error << std::endl;
        // DO NOT RETURN
    }
```

This example compiles and run, but when I for example remove comma after "model" I get SIGSEGV while accessing "tire_pressure" array.

Thanks for help!

**Comment by jkeiser** [comment]
You should be able to continue as long as the json *structure* is not compromised. In the example of a missing comma, continuing will yield problematic results.

String vs int is fine. You can check the error code, I think it's INCORRECT_TYPE or something similar (see documentation on errors).

**Reply by kgs** [reply]
Yeah, I am aware that there are different error codes, but in some parts in documentation there is information that user should not rely on it (because they may change in future releases). So can I rely on them or not? :)

**Comment by jkeiser** [comment]
You should be able to rely on type errors (errors that occur when you assume a json value has a certain type, range, or field): INCORRECT_TYPE, INDEX_OUT_OF_BOUNDS, and NO_SUCH_FIELD. NUMBER_OUT_OF_RANGE feels like a grey area to me: I'd have to review the code but it *ought* to be something you can rely on as well.

Anything else is not guaranteed simply because usually when there is one error there are multiple problems you could diagnose it as.

We need to update our documentation for this; I don't see any notes in basics.md. Thanks!

**Comment by jkeiser** [comment]
The best answer here is probably that in general, you cannot continue parsing in the face of an error. Type errors are really designed as ways to test the type of a value, and it's intended that if you check it, you try it as another type until you achieve success. Like if (incorrect type) try as double; if (incorrect type) exit.

Apart from "retry this value as something else," continuing on error is not a well tested scenario at all.

**Comment by lemire** [comment, marked as answer]
I recommend using `type()` to determine the type of a value. The method returns a json_type value describing the value (json_type::array, json_type::object, json_type::number, json_type::string, json_type::boolean, json_type::null). It is more convenient than trying something and catching an error, and then trying something else. 

Regarding your JSON document, I concur with @jkeiser : if the structure is not that of a JSON document at all, then continuing after receiving an error can probably lead to a bad condition.

What we do support, and I had to use it recently for a project, is that if you have a value that is invalid (say a string that contains bad escaped characters), then you can bypass the get_string method and use raw_json_token to process the string yourself. The same applies to numbers... you don't have to rely on our number parser, you can just grab a strint_view of the value and do whatever you like with it. In that sense, we do support 'bad JSON' inputs, as long as the structure is intact.

This should be better documented and explained, ideally with examples.

There are limits to what we will accept, however... so we don't accept non UTF-8 inputs, for example.
